### PR TITLE
[SYMFONY62] OhMySMTP, ArgumentValueResolverInterface, Security etc. #278

### PR DESCRIPTION
**Pull request by JohJohan**

### CODE REVIEW

**Comment by johanadivare** [comment]
Fixes https://github.com/rectorphp/rector-symfony/issues/278

**Comment by johanadivare** [comment]
What is the best way to remove the function call to `->createView()`? I can make a rule in class method and check if it has a render function call that contains `form`?

**Comment by stefantalen** [comment]
1. Check arguments of `AbstractController::render()`
2. Check if argument is an instance of `FormInstance`
3. Remove `->createView()` from argument

However, doing it this way might not update implementations like:
```php
$form = $contactForm->createView();
return $this->render('template.html.twig', [
    contact_form => $form,
];
```
But I think it'll be safer to limit this feature to update calls within the `render()` method only.

**Comment by johanadivare** [comment]
@stefantalen i agree, and the implementation like you mention also still works

**Comment by TomasVotruba** [comment]
Let me know when this is ready.

The createView() removal will indeed need a custom rule.  The conditions list seem valid.

**Review by johanadivare — APPROVED** [review]

**Comment by samsonasik** [comment]
Please run:

```
vendor/bin/rector && composer fix-cs 
```

to make CI green.

File: src/Rector/MethodCall/SimplifyFormRenderingRector.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (! $this->isObjectType(
+            $node->var,
+            new ObjectType('Symfony\Bundle\FrameworkBundle\Controller\AbstractController')
+        )) {
+            return null;
+        }
+
```
Comment (samsonasik) [inline src/Rector/MethodCall/SimplifyFormRenderingRector.php:74]: first class callable should be skipped:

```suggestion
        if ($node->isFirstClassCallable()) {
            return null;
        }
```

Comment (johanadivare) [reply src/Rector/MethodCall/SimplifyFormRenderingRector.php:75]: ++

File: src/Rector/MethodCall/SimplifyFormRenderingRector.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            }
+            $replaced = true;
+            $arrayItem->value = $arrayItem->value->var;
+        }
+
+        if (! $replaced) {
+            return null;
+        }
+
+        return $node;
```
Comment (samsonasik) [inline src/Rector/MethodCall/SimplifyFormRenderingRector.php:115]: You can move this logic to separate method to make PHPStan green, eg: `

```php
/**
 * @param ArrayItem[]|null[] $items
 */
private function processRemoveCreateView(array $items): ?MethodCall
```

**Comment by samsonasik** [comment]
You can verify phpstan with run:

```
composer phpstan
```

File: src/Rector/MethodCall/SimplifyFormRenderingRector.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            return null;
+        }
+
+        if (! isset($node->args[1])) {
+            return null;
+        }
+
+        if (! $node->args[1] instanceof Arg) {
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (johanadivare) [reply src/Rector/MethodCall/SimplifyFormRenderingRector.php:90]: I can check isset on value?

[the remark this answers is not in the document]
Comment (samsonasik) [reply src/Rector/MethodCall/SimplifyFormRenderingRector.php:90]: I see, assign to variable with `@var Arg $variableName` should be ok

File: src/Rector/MethodCall/SimplifyFormRenderingRector.php (hunk 2 above)
Comment (johanadivare) [reply src/Rector/MethodCall/SimplifyFormRenderingRector.php:115]: ++

File: src/Rector/MethodCall/SimplifyFormRenderingRector.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (johanadivare) [reply src/Rector/MethodCall/SimplifyFormRenderingRector.php]: ++

**Comment by johanadivare** [comment]
@samsonasik its ready now :)

**Comment by johanadivare** [comment]
@TomasVotruba its ready

**Comment by johanadivare** [comment]
Note i added a bit more rules

File: config/sets/symfony/symfony62.php [hunk 4]
```diff
@@ -16,6 +23,53 @@
...
+            // @see https://github.com/symfony/symfony/pull/45034
+            new MethodCallRename(
+                'Symfony\Component\HttpFoundation\Request',
+                'getContentType',
+                'getContentTypeFormat'
+            ),
+            // @see https://github.com/symfony/symfony/pull/47711
+            new MethodCallRename('Symfony\Component\Mime\Email', 'attachPart', 'addPart'),
+            // @see https://github.com/symfony/symfony/pull/47363
+            new MethodCallRename(
```
Comment (johanadivare) [inline config/sets/symfony/symfony62.php:60]: Is this allowed because return type is diffrent supports: bool resolve: array

**Comment by TomasVotruba** [comment]
Thank you :clap:

File: config/sets/symfony/symfony62.php (hunk 4 above)
Comment (samsonasik) [reply config/sets/symfony/symfony62.php]: Per comment on https://github.com/rectorphp/rector-symfony/pull/283#discussion_r1029286911

this part may  need to be reverted

**Pull request merged**